src/mem3_sync_event_listener.erl

```diff
@@ -51,35 +68,136 @@ handle_event(NodesDb, updated, #state{nodes = NodesDb} = St) ->
-    {ok, St};
+    maybe_push_shards(St);
 handle_event(<<"shards/", _/binary>> = ShardName, updated, St) ->
-    try mem3:shards(mem3:dbname(ShardName)) of
-    Shards ->
-        Targets = [S || #shard{node=N, name=Name} = S <- Shards,
-            N =/= node(), Name =:= ShardName],
-        Live = nodes(),
-        [mem3_sync:push(ShardName,N) || #shard{node=N} <- Targets,
-            lists:member(N, Live)]
-    catch error:database_does_not_exist ->
-        ok
-    end,
-    {ok, St};
+    Buckets = bucket_shard(ShardName, St#state.buckets),
+    maybe_push_shards(St#state{buckets=Buckets});
 handle_event(<<"shards/", _:18/binary, _/binary>> = ShardName, deleted, St) ->
     mem3_sync:remove_shard(ShardName),
-    {ok, St};
+    maybe_push_shards(St);
 handle_event(_DbName, _Event, St) ->
-    {ok, St}.
+    maybe_push_shards(St).
 
+handle_cast({set_frequency, Frequency}, St) ->
+    #state{delay = Delay, buckets = Buckets0} = St,
+    Buckets1 = rebucket_shards(Delay, Frequency, Buckets0),
+    maybe_push_shards(St#state{frequency=Frequency, buckets=Buckets1});
+handle_cast({set_delay, Delay}, St) ->
+    #state{frequency = Frequency, buckets = Buckets0} = St,
+    Buckets1 = rebucket_shards(Delay, Frequency, Buckets0),
+    maybe_push_shards(St#state{delay=Delay, buckets=Buckets1});
 handle_cast(Msg, St) ->
     couch_log:notice("unexpected cast to mem3_sync_event_listener: ~p", [Msg]),
-    {ok, St}.
+    maybe_push_shards(St).
 
+handle_info(timeout, St) ->
+    maybe_push_shards(St);
 handle_info(Msg, St) ->
     couch_log:notice("unexpected info to mem3_sync_event_listener: ~p", [Msg]),
+    maybe_push_shards(St).
+
+handle_config_change("mem3", "sync_delay", Delay0, _, St) ->
+    try list_to_integer(Delay0) of
+        Delay1 ->
+            couch_event_listener:cast(
+                ?MODULE,
+                {set_delay, Delay1}
+            )
+    catch error:badarg ->
+        couch_log:warning(
+            "ignoring bad value for mem3.sync_delay: ~p",
+            [Delay0]
+        )
+    end,
+    {ok, St};
+handle_config_change("mem3", "sync_frequency", Frequency0, _, St) ->
+    try list_to_integer(Frequency0) of
+        Frequency1 ->
+            couch_event_listener:cast(
+                ?MODULE,
+                {set_frequency, Frequency1}
+            )
+    catch error:badarg ->
+        couch_log:warning(
+            "ignoring bad value for mem3.sync_frequency: ~p",
+            [Frequency0]
+        )
+    end,
+    {ok, St};
+handle_config_change(_, _, _, _, St) ->
     {ok, St}.
+
+handle_config_terminate(_, stop, _) -> ok;
+handle_config_terminate(_Server, _Reason, St) ->
+    Fun = fun() ->
+        timer:sleep(5000),
+        config:listen_for_changes(?MODULE, St)
+    end,
+    spawn(Fun).
+
+bucket_shard(ShardName, [B|Bs]=Buckets0) ->
+    case waiting(ShardName, Buckets0) of
+        true -> Buckets0;
+        false -> [sets:add_element(ShardName, B)|Bs]
+    end.
```

Comment:
In the high throughput case this algorithm will generally result in 9 `is_element` checks that return false and a final check that returns true, right? It seems like it would be more efficient if the list of sets was stored with the oldest bucket at the head.